trl/trainer/kto_trainer.py

```diff
@@ -0,0 +1,941 @@
+                reference_logp = self.compute_reference_log_probs(padded_batch)
+                reference_logp = self.accelerator.gather_for_metrics(reference_logp)
+                reference_logps.append(reference_logp.cpu())
+
+            all_reference_logps = torch.cat(reference_logps).float().numpy()
+            eval_dataset = eval_dataset.add_column(name="reference_logps", column=all_reference_logps)
+
+            # Save calculated reference_chosen_logps and reference_rejected_logps to the eval_dataset for subsequent runs
+            if self.eval_dataset is not None:
+                self.eval_dataset = eval_dataset
+            self._precomputed_eval_ref_log_probs = True
+
+        return super().get_eval_dataloader(eval_dataset=eval_dataset)
+
+    def compute_reference_log_probs(self, padded_batch: Dict) -> Dict:
+        """Computes log probabilities of the reference model for a single padded batch of a KTO specific dataset."""
+        with torch.no_grad():
+            if self.ref_model is None:
+                with self.accelerator.unwrap_model(
+                    self.model
+                ).disable_adapter() if self.is_peft_model else nullcontext():
+                    if self.is_encoder_decoder:
+                        all_logits = self.model(
+                            padded_batch["prompt_input_ids"],
+                            attention_mask=padded_batch["prompt_attention_mask"],
+                            decoder_input_ids=padded_batch["generation_decoder_input_ids"],
+                            labels=padded_batch["generation_labels"],
+                        ).logits
+                    else:
+                        all_logits = self.model(
+                            padded_batch["generation_input_ids"],
+                            attention_mask=padded_batch["generation_attention_mask"],
+                        ).logits
+            else:
+                if self.is_encoder_decoder:
+                    all_logits = self.ref_model(
+                        padded_batch["prompt_input_ids"],
+                        attention_mask=padded_batch["prompt_attention_mask"],
+                        decoder_input_ids=padded_batch["generation_decoder_input_ids"],
+                        labels=padded_batch["generation_labels"],
+                    ).logits
+                else:
+                    all_logits = self.ref_model(
+                        padded_batch["generation_input_ids"], attention_mask=padded_batch["generation_attention_mask"]
+                    ).logits
+
+        return self.get_batch_logps(
+            all_logits,
+            padded_batch["generation_labels"],
+            average_log_prob=False,
+            is_encoder_decoder=self.is_encoder_decoder,
+            label_pad_token_id=self.label_pad_token_id,
+        )
+
+    def build_tokenized_answer(self, prompt, answer):
+        """
+        Llama tokenizer does satisfy `enc(a + b) = enc(a) + enc(b)`.
```

Comment:
Should this be "not"?
```suggestion
        Llama tokenizer does not satisfy `enc(a + b) = enc(a) + enc(b)`.
```